letsencrypt-auto-source/pieces/bootstrappers/rpm_python3_legacy.sh

```diff
@@ -0,0 +1,73 @@
+# If new packages are installed by BootstrapRpmPython3 below, this version
+# number must be increased.
+BOOTSTRAP_RPM_PYTHON3_LEGACY_VERSION=1
+
+# Checks if rh-python36 can be installed.
+Python36SclIsAvailable() {
+  InitializeRPMCommonBase >/dev/null 2>&1;
+
+  if "${TOOL}" list rh-python36 >/dev/null 2>&1; then
+    return 0
+  fi
+  if "${TOOL}" list centos-release-scl >/dev/null 2>&1; then
+    return 0
+  fi
+  return 1
+}
+
+# Try to enable rh-python36 from SCL if it is necessary and possible.
+EnablePython36SCL() {
+  if "$EXISTS" python3.6 > /dev/null 2> /dev/null; then
+      return 0
+  fi
+  if [ ! -f /opt/rh/rh-python36/enable ]; then
+      return 0
+  fi
+  . /opt/rh/rh-python36/enable
```

Comment:
I am strongly against that. The reason is that in general, continuing a script when a command returned a non 0 exit code is a bad idea, because something wrong happened and most likely the system is not in the state expected. Here we would ignore all errors on bad behaviors that can happen (broken enabler script, broken shell ...) and continue the logic without a functional Python distribution, only to catch a hypothetical non 0 exit code in the future.

We called `set +e` before because `scl_source` was relying on this, but this behavior from `scl_source` was already a questionable design decision. But since it was required, we had to adapt ourselves.

However here, nothing is saying that `/opt/rh/rh-python36/enable` could ever return a non 0 exit code for a good reason, and I cannot see how changing this behavior could be a good idea. Indeed if maintainers decided to do so, all scripts in the wild that use `rh-python36` and enable it would become unstable.

Reply:
Keep in mind that there's a big difference between:

1. Using `set` to modify the shell's default behavior and sourcing a script.
2. Ignoring that script's exit code.

The former causes the script to execute in your modified environment changing the default behavior of the shell interpreting the script while the latter treats the script like a black box and only worries about the exit code the script returns to the caller.

By using `/opt/rh/rh-python36/enable` directly, I think we're reaching into the internals of the package therefore I don't think we should rely on having consistent behavior. Instead, I think Red Hat wants you to use `scl` (or in very rare cases `scl_source`) to interface with these scripts.

As a test, I added the following to the top of `/opt/rh/rh-python36/enable` and both `scl` and `scl_source` still work:
```
try_something
if [ "$?" -ne 0 ]; then
  do_fallback
fi
```
If they did the equivalent of `set -e; . /opt/rh/rh-python36/enable` it would not.

Since the official scripts do not do this, I do not think it is worth doing it ourselves as I think it makes certbot-auto too fragile to changes I suspect the SCL maintainers view as internal.

With that said, I think checking the exit code of `. /opt/rh/rh-python36/enable` is a good idea. Both `scl` and `scl_source` does do this so I'd prefer this code looks something like:
```
set +e
if ! . /opt/rh/rh-python36/enable; then
  error 'Unable to enable rh-python36!'
  exit 1
fi
set -e
```

Reply:
OK, I see your point. Indeed I forgot the difference between sourcing a script and calling it. Your proposition matches my concern about controling the sourcing result. I implement that.

Reply:
Done.